src/main/java/com/epam/aidial/core/storage/ResourceDescription.java

```diff
@@ -60,31 +66,56 @@ public String getParentPath() {
         return parentFolders == null ? null : String.join(BlobStorageUtil.PATH_SEPARATOR, parentFolders);
     }
 
-    public static ResourceDescription from(ResourceType type, String bucketName, String bucketLocation, String relativeFilePath) {
+    /**
+     * Creates resource for the given parameters
+     *
+     * @param type           resource type
+     * @param bucketName     bucket name (encrypted)
+     * @param bucketLocation bucket location on blob storage; bucket location must end with /
+     * @param path           url encoded relative path; url path is null or empty we treat it as user home
+     */
+    public static ResourceDescription from(ResourceType type, String bucketName, String bucketLocation, String path) {
         // in case empty path - treat it as a home folder
-        relativeFilePath = StringUtils.isBlank(relativeFilePath) ? BlobStorageUtil.PATH_SEPARATOR : relativeFilePath;
+        String urlEncodedRelativePath = StringUtils.isBlank(path) ? BlobStorageUtil.PATH_SEPARATOR : path;
         verify(bucketLocation.endsWith(BlobStorageUtil.PATH_SEPARATOR), "Bucket location must end with /");
 
-        String[] elements = relativeFilePath.split(BlobStorageUtil.PATH_SEPARATOR);
+        String[] encodedElements = urlEncodedRelativePath.split(BlobStorageUtil.PATH_SEPARATOR);
+        List<String> elements = Arrays.stream(encodedElements).map(ResourceDescription::urlDecode).toList();
+        elements.forEach(element ->
+                verify(isValidFilename(element), "Invalid path provided " + urlEncodedRelativePath)
+        );
         List<String> parentFolders = null;
         String name = "/";
-        if (elements.length > 0) {
-            name = elements[elements.length - 1];
+        if (!elements.isEmpty()) {
+            name = elements.get(elements.size() - 1);
         }
-        if (elements.length > 1) {
-            String parentPath = relativeFilePath.substring(0, relativeFilePath.length() - name.length() - 1);
+        if (elements.size() > 1) {
+            String parentPath = urlEncodedRelativePath.substring(0, urlEncodedRelativePath.length() - name.length() - 1);
```

Comment:
Name is decoded segment which can have different size. Add an unit test for parsing.